src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_black_knight.cpp

```diff
@@ -15,337 +15,470 @@
+            if (Creature* knight = instance->GetCreature(DATA_BLACK_KNIGHT))
             {
-                if (uiResurrectTimer <= uiDiff)
-                {
-                    me->SetFullHealth();
-                    DoCast(me, SPELL_BLACK_KNIGHT_RES, true);
-                    uiPhase++;
-                    uiResurrectTimer = 4000;
-                    bEventInProgress = false;
-                    me->ClearUnitState(UNIT_STATE_ROOT | UNIT_STATE_STUNNED);
-                } else uiResurrectTimer -= uiDiff;
+                if (!knight->HasAura(SPELL_BLACK_KNIGHT_DIE))
+                    knight->AI()->KilledUnit(who);
             }
+        }
+
+        void UpdateAI(uint32 uiDiff) override
+        {
+            events.Update(uiDiff);
 
-            switch (uiPhase)
+            if (me->HasUnitState(UNIT_STATE_CASTING))
+                return;
+
+            if (!UpdateVictim())
+                return;
+
+            while (uint32 eventId = events.ExecuteEvent())
             {
-                case PHASE_UNDEAD:
-                case PHASE_SKELETON:
+                switch (eventId)
                 {
-                    if (uiIcyTouchTimer <= uiDiff)
-                    {
-                        DoCastVictim(SPELL_ICY_TOUCH);
-                        uiIcyTouchTimer = urand(5000, 7000);
-                    } else uiIcyTouchTimer -= uiDiff;
-                    if (uiPlagueStrikeTimer <= uiDiff)
-                    {
-                        DoCastVictim(SPELL_ICY_TOUCH);
-                        uiPlagueStrikeTimer = urand(12000, 15000);
-                    } else uiPlagueStrikeTimer -= uiDiff;
-                    if (uiObliterateTimer <= uiDiff)
-                    {
-                        DoCastVictim(SPELL_OBLITERATE);
-                        uiObliterateTimer = urand(17000, 19000);
-                    } else uiObliterateTimer -= uiDiff;
-                    switch (uiPhase)
-                    {
-                        case PHASE_UNDEAD:
+                    case EVENT_CLAW:
+                        DoCastVictim(SPELL_CLAW);
+                        if (Unit* target = SelectTarget(SELECT_TARGET_RANDOM, 1, 50.0f, true))
                         {
-                            if (uiDeathRespiteTimer <= uiDiff)
-                            {
-                                if (Unit* target = SelectTarget(SELECT_TARGET_RANDOM, 0, 100, true))
-                                {
-                                    if (target->IsAlive())
-                                        DoCast(target, SPELL_DEATH_RESPITE);
-                                }
-                                uiDeathRespiteTimer = urand(15000, 16000);
-                            } else uiDeathRespiteTimer -= uiDiff;
-                            break;
+                            ResetThreatList();
+                            AddThreat(target, 10.0f);
+                            me->AI()->AttackStart(target);
```

Comment:
`me->AI()` is `this`

also I don't like manual `AttackStart` calls here, `UpdateVictim` will do this anyway